File: Mavim-services/Topic/Mavim.Manager.Api.Topic.Repository/v1/Mappers/Abstract/GenericHyperlinkFieldMapper.cs
using Microsoft.Extensions.Logging;
using System;
using IModel = Mavim.Manager.Model;
using IRepo = Mavim.Manager.Api.Topic.Repository.Interfaces.v1.Fields;

namespace Mavim.Manager.Api.Topic.Repository.v1.Mappers.Abstract
{
    internal abstract class GenericHyperlinkFieldMapper<TModel, TRepo> : GenericFieldMapper<TModel, TRepo> where TModel : IModel.ISimpleField where TRepo : IRepo.IField
    {
        private readonly ILogger _logger;

        public GenericHyperlinkFieldMapper(ILoggerFactory loggerFactory)
        {
            _logger = loggerFactory?.CreateLogger(GetType().Name) ?? throw new ArgumentNullException(nameof(loggerFactory));
        }

        protected Uri Map(object fieldValue)
        {
            if (Uri.TryCreate(fieldValue?.ToString(), UriKind.Absolute, out Uri result))
                return result;

            _logger.LogWarning($"Unable to map fieldvalue to URL.");
            return null;
        }
    }
}

File: Mavim-apps/Mavim.frontend/projects/mavim-app/src/shared/field/mappers/abstract/generic-hyperlink-field.mapper.ts
import { GenericFieldMapper } from './generic-field.mapper';

export abstract class GenericHyperlinkFieldMapper<
	T
> extends GenericFieldMapper<T> {
	protected mapToString(url: URL): string {
		return url?.toString() || '';
	}

	protected mapToUrl(url: string): URL {
		if (!url) {
			return undefined;
		}
		return new URL(url);
	}
}
